src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
```

Comment:
> I don't want to make this more complex without evidence that contention is hurting us.

Contention is pretty continually hurting me. The question is less whether or not things need to be made concurrent, and more when they will show up on top of my list because so many places in the code currently block. :smile: